### DISCUSSION TITLE
Conditional 2FA?

### DISCUSSION THREAD

**Discussion by Leonardo-Ferreira** [keycloak/keycloak#29243] [Q&A] [answered]
I would like that Keycloak, after validating user/password would call my security api and if that api returns a positive result, Keycloak would ask the user for a OTP that he can choose if he will use email, totp or sms...

how can I do that?

**Comment by codespearhead** [comment]
You'll have to create a custom Keycloak Service Provider Interface (SPI) to call your API, similar to the one in #29232 .

Here's a [demo](https://github.com/dasniko/keycloak-2fa-sms-authenticator/tree/main/src/main/java/dasniko/keycloak/authenticator).

**Comment by Leonardo-Ferreira** [comment]
Thank you for the reference, but I couldn't quite figure out how am I supposed to "skip" 2FA if the API returns "safe"? can you clarify?

**Reply by codespearhead** [reply, marked as answer]
You can do that via a Conditional Authenticator.

Here's a concise but slightly outdated [tutorial](https://medium.com/@rishabhsvats/keycloak-custom-conditional-authenticator-5e50d61f8d12), and here's [a more recent one](https://karelics.fi/building-a-custom-conditional-authenticator-for-keycloak/).

**Reply by codespearhead** [reply]
@Leonardo-Ferreira Does that answer your question? If so, can you [mark this discussion as answered](https://docs.github.com/en/discussions/managing-discussions-for-your-community/moderating-discussions#marking-a-comment-as-an-answer) so it doesn't show up in the discussion tab as unanswered?
